retis/src/helpers/bimap.rs

```diff
@@ -0,0 +1,148 @@
+}
+
+impl<'a, K: Ord, V> KeyRange<'a, K, V> {
+    fn new(map: &'a BTreeMap<Arc<K>, Arc<V>>, range: (Bound<&'a K>, Bound<&'a K>)) -> Self {
+        Self {
+            inner: map.range::<K, (Bound<&K>, Bound<&K>)>(range),
+        }
+    }
+}
+
+impl<'a, K: Ord, V> Iterator for KeyRange<'a, K, V> {
+    type Item = (&'a K, &'a V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next().map(|(k, v)| (k.as_ref(), v.as_ref()))
+    }
+}
+
+impl<'a, K: Ord, V> DoubleEndedIterator for KeyRange<'a, K, V> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        self.inner
+            .next_back()
+            .map(|(k, v)| (k.as_ref(), v.as_ref()))
+    }
+}
+
+/// A bimap backed by two `BTreeMap`.
+pub(crate) struct BiBTreeMap<K, V> {
+    /// A `BTreeMap` of `Arc` wrapped keys and values representing the key->value
+    /// relationship.
+    left2right: BTreeMap<Arc<K>, Arc<V>>,
+    /// A `BTreeMap` of `Arc` wrapped keys and values representing the value->key
+    /// relationship.
+    right2left: BTreeMap<Arc<V>, Arc<K>>,
+}
+
+impl<K, V> BiBTreeMap<K, V>
+where
+    K: std::hash::Hash + Eq + Ord + std::fmt::Display,
+    V: std::hash::Hash + Eq + Ord,
+{
+    /// Crate an empty `BiBTreeMap`.
+    pub(crate) fn new() -> Self {
+        BiBTreeMap {
+            left2right: BTreeMap::new(),
+            right2left: BTreeMap::new(),
+        }
+    }
+
+    /// Inserts the given key->value pair into the bimap.
+    pub(crate) fn insert(&mut self, key: K, value: V) {
+        let rc_key = Arc::new(key);
+        let rc_value = Arc::new(value);
+        self.right2left.insert(rc_value.clone(), rc_key.clone());
+        self.left2right.insert(rc_key, rc_value);
+    }
+
+    /// Returns a reference to the value corresponding to the given key.
+    pub(crate) fn get_by_left(&self, key: &K) -> Option<&V> {
+        self.left2right.get(key).map(|x| x.as_ref())
+    }
+
+    /// Returns a reference to the key corresponding to the given value.
+    pub(crate) fn get_by_right(&self, key: &V) -> Option<&K> {
+        self.right2left.get(key).map(|x| x.as_ref())
+    }
+
+    /// Creates an iterator over the key->value pairs within a range of keys in the bimap in
+    /// ascending order.
+    pub(crate) fn range_by_left<'a, R>(&'a self, target: &'a R) -> KeyRange<'_, K, V>
+    where
+        Arc<K>: Borrow<K>,
+        R: RangeBounds<K>,
+    {
+        let start = target.start_bound();
+        let end = target.end_bound();
+        KeyRange::new(&self.left2right, (start, end))
+    }
+
+    /// Returns the number of key->value pairs in the bimap.
+    pub(crate) fn len(&self) -> usize {
+        self.left2right.len()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::BiBTreeMap;
+    use std::ops::Bound::{Included, Unbounded};
+
+    fn bimap_init() -> BiBTreeMap<i32, i32> {
+        let mut bimap = BiBTreeMap::new();
+        bimap.insert(1, 2);
+        bimap.insert(3, 4);
+        bimap.insert(5, 6);
```

Comment:
Nit: can we add a key not in the range of the `range_left` test here? Eg. `(9, 8)`.